### ISSUE TITLE
Come up with a good user API

### ISSUE THREAD

**Issue by contsili** [predictive-clinical-neuroscience/PCNtoolkit#456]
### API suggestion 1
user has to run both `model.compute_correlation_matrix` and `model.compute_thrivelines` after they have fitted their data

**Pros**

Can specify all options for compute_correlation_matrix and compute_thrivelines
Clear call separation between longitudinal and cross-sectional data
Easy access to the correlation matrix (`corrmat`)

**Cons**
Less user friendly
Less dev friendly

---
### API suggestion 2
`model = NormativeModel(longitudinal=True)`

**Pros**
very user friendly and simple

**Cons**
Harder to expose all options for compute_correlation_matrix and compute_thrivelines
unclear how corrmat should be exposed
unclear call separation between longitudinal and cross-sectional data

---
### API suggestion 3
Make a `LongitudinalModel `class which will define all the input arguments for model.compute_correlation_matrix and model.compute_thrivelines and also contain these functions. Then this class will be called by NormativeModel(BLR(), LongitudinalModel()...). 

**Pros** 
user friendly and dev friendly (extendable, maintanable)
Can specify all options for compute_correlation_matrix and compute_thrivelines

**Cons**
unclear how corrmat should be exposed
unclear call separation between longitudinal and cross-sectional data

**Comment by contsili** [comment]
### LNM pipeline:

1. **Data**
- LNM data (test set) - site1, site2, site3. 
- LNM data can also be in the train set but they will be treated as cross-sectional data during training (sub1-visit1 and sub1-visit2 are treated as two separate data points)
- Cross sectional data (training, test) - site1, site2. About the site3 from the LNM data, you either remove it, or find a subj from site3 for the cross sectional data , or transfer (not recommended))
2. **train_test_split**
3. **BLR/HBR**
4. **NormativeModel().fit_predict**
- Z scores and centiles computed for test data (as we normally do in a NM workflow)
5. **LNM step**
- test data: needs 6 more columns:` site` `sex ``age `(to calculate the correllation between ages) `subjID ``visitID `(to know what are the LNM data) `disease_status `
- Remove cross-sectional data from test
- compute correlation matrix. The input to the correlation matrix has to be the LNM test data (not train)


Note there can be a transfer with LNM data: eg `longitudinal_model.transfer_LNM(corrmat of pretrained LNM data, local LNM)`

### LNM API suggestion:
For step 5 make a new class called LongitudinalModel(), this will have methods:
LongitudinalModel().load_corr_mat -> corr_mat
LongitudinalModel().save_corr_mat
LongitudinalModel().compute_corr_mat -> corr_mat

LongitudinalModel().compute_thrivelines -> NormData with thriveline info

LongitudinalModel().compute_zgain
LongitudinalModel().compute_zdiff (works only with BLR - write a warning/error for that)

### More ideas
in `NormData `add a function `remove_cross_sectional_data`
`NormativeModel().fit_predict` can keep more info in the NormData object ` site` `sex ``age ` `subjID ``visitID ``disease_status `

**Comment by contsili** [comment]
This is an example of how we expect the users to run their analysis after we implement the Longitudinal Modelling:
 
We will support two methods:
 
- **z-diff** — change between two visits. Two timepoints only.
- **z-gain** — generalised velocity centile. Any number of timepoints. Also gives
  thrive lines (overlaid on centile plots) and conditional forecasts.

Both methods require a fitted `NormativeModel` and longitudinal data (cross sectional data shoould be removed from the NormData object after you have fitted your NM earlier in your analysis), with z-scores already predicted.
 
---
 
# Step 0 — prepare longitudinal data from a NormData (future implementation)
 
Your `NormData` may contain a mix of single-visit and multi-visit subjects. Extract
the longitudinal subset before doing Longitudinal Normative Modelling (LNM):
 
```python
# test_data is a NormData with both single- and multi-visit subjects
controls_long = test_data.select_longitudinal(
    subject_id_col='sub_id',
) # visits can be encoded in a visit column eg 1,2,3 or we can also have multiple columns (phenotype x visit)
```
 
# step 1 - LNM scores
There are two ways to compute the scores that show if your subject has an unexpected (non healthy) longitudinal trajectory:
 
## z-diff (Rehák Bučková 2025)
 
**Two timepoints only. BLR or wBLR models only.**
 
```python
from pcntoolkit.longitudinal import ZDiffScore
 
zdiff = ZDiffScore(
    normative_model=normative_model,
    norm_data=test_data, 
    subject_id='sub_id',
)
 
# Score patients — returns DataArray (subjects × response_vars)
# patients_data is a NormData with both visits and z-scores predicted
scores = zdiff.score(test_data, subject_id='sub_id', timepoint_col='visit')
```
 
`ZDiffScore` internally estimates σ²(1-ρ) per response variable  caches it.
 
 ----
Another implementation to consider is keeping in the data the `disease_status` e.g., control vs patient and computing the zdiff score for the patients, see example from the [protocol update 25 paper](https://github.com/predictive-clinical-neuroscience/pu25_code/blob/main/notebooks/secondary_workflows/longitudinal_modelling.ipynb).

The code then becomes:

```python
from pcntoolkit.longitudinal import ZDiffScore
 
zdiff = ZDiffScore(
    normative_model=normative_model,
    norm_data=control_data, # σ²(1-ρ) should be computed on the controls
    subject_id='sub_id',
)
 
# Score patients — returns DataArray (subjects × response_vars)
# patients_data is a NormData with both visits and z-scores predicted
scores = zdiff.score(patient_data, subject_id='sub_id', timepoint_col='visit')
```


## z-gain (Bayer et al. 2026)
 
**Any number of timepoints. Any RegressionModel.**
 
```python
from pcntoolkit.longitudinal import ZGainScore
 
zgain = ZGainScore(
    normative_model=fitted_model,
    norm_data=test_data, 
    subject_id='sub_id',
    bandwidth=5,         # smoothing bandwidth for R estimation
    covariate='age',     # covariate that indexes the correlation matrix
)
 
# Score patients — NormData with all visits, z-scores predicted
scores = zgain.score(patients_data, subject_id='sub_id', timepoint_col='visit')
 
# Conditional forecast: predict the next timepoint given a history. This is future implementation.
mean, std = zgain.conditional_forecast(
    z_history=[z_t1, z_t2, z_t3],
    ages_history=[55.0, 56.2, 57.1],
    target_age=58.0,
)
```

 
### Accessing the correlation matrix
 
After any call that triggers R estimation, it is available as a computed attribute:
 
```python
# DataArray indexed by (age_t1, age_t2, response_var)
R = zgain.correlation_matrix
 
# Example: correlation between age 30 and age 32 for one region
r = R.sel(response_var='lh_G_front_middle_thickness').values

# you can then plot the R
```
 
# step 3 - plotting
 
## Plotting thrive lines on centile plots
 
Thrive lines are overlaid directly on centile plots via `plot_centiles_advanced`.
 
```python
from pcntoolkit.plotter import plot_centiles_advanced
 
# Without thrive lines (existing behaviour, unchanged)
plot_centiles_advanced(
    normative_model=fitted_model,
    data=test_data,
    response_var='lh_G_front_middle_thickness',
)
 
# With thrive lines overlaid
plot_centiles_advanced(
    normative_model=fitted_model,
    data=test_data,
    response_var='lh_G_front_middle_thickness',
    
    # thrivelines
    thrive=zgain,             # pass the ZGainScore instance
    thrive_span=5,            # interval in years
    thrive_z=1.96,            # significance threshold
)
```
 
When `thrive` is provided, `plot_centiles_advanced` calls `zgain.thrive_lines(...)`
internally and overlays the result on the existing centile plot. All existing
`plot_centiles_advanced` kwargs work as before.
 
---
 
## Choosing between z-diff and z-gain
 
| You have…                              | Use          |
|----------------------------------------|--------------|
| Two timepoints, BLR model              | `ZDiffScore` — simpler, scalar variance estimate |
| Two timepoints, want thrive lines      | `ZGainScore` |
| Three or more timepoints               | `ZGainScore` (z-diff doesn't generalise) |
| Want to forecast future visits         | `ZGainScore.conditional_forecast()` |
